python/metatensor_core/metatensor/data/array.py

```diff
@@ -158,77 +158,57 @@ class DeviceWarning(RuntimeWarning):
 
 
 class ArrayWrapper:
-    """Small wrapper making Python arrays compatible with ``mts_array_t``."""
+    """Store together some Python array and it's shape as C-compatible data"""
 
-    def __init__(self, array):
-        self.array = array
-        self._shape = ctypes.ARRAY(c_uintptr_t, len(array.shape))(*array.shape)
-
-        if _is_numpy_array(array):
-            array_origin = _origin_numpy()
-        elif _is_torch_array(array):
-            array_origin = _origin_pytorch()
-        else:
-            raise ValueError(f"unknown array type: {type(array)}")
-
-        mts_array = mts_array_t()
-        # `mts_array_t::ptr` is a pointer to the PyObject `self`
-        mts_array.ptr = ctypes.cast(
-            ctypes.pointer(self._get_py_object()), ctypes.c_void_p
-        )
-
-        @catch_exceptions
-        def mts_array_origin(this, origin):
-            origin[0] = array_origin
+    __slots__ = ["array", "c_shape"]
 
-        # use storage.XXX.__class__ to get the right type for all functions
-        mts_array.origin = mts_array.origin.__class__(mts_array_origin)
+    def __init__(self, array, c_shape):
+        self.array = array
+        self.c_shape = c_shape
 
-        mts_array.data = mts_array.data.__class__(_mts_array_data)
 
-        mts_array.shape = mts_array.shape.__class__(_mts_array_shape)
-        mts_array.reshape = mts_array.reshape.__class__(_mts_array_reshape)
-        mts_array.swap_axes = mts_array.swap_axes.__class__(_mts_array_swap_axes)
+# We keep wrappers of Python arrays alive by storing a reference to them in this
+# dictionary. Each value is stored under the `id(value)` key, which makes sure no two
+# values have the same key.
+_KNOWN_ARRAY_WRAPPERS = {}
 
-        mts_array.create = mts_array.create.__class__(_mts_array_create)
-        mts_array.copy = mts_array.copy.__class__(_mts_array_copy)
-        mts_array.destroy = mts_array.destroy.__class__(_mts_array_destroy)
+# cache array types by array length, since creating the type everytime is a bit slow
+_POSSIBLE_C_SHAPE_TYPES = [ctypes.ARRAY(c_uintptr_t, N) for N in range(20)]
```

Comment:
This means that we don't support more dimensions than 20?

Reply:
Perhaps should be changed to 32 to match numpy

Reply:
Yes, good point! Do you know if torch also has a limit on Tensor rank?